Simplify eo3 detection: use only $schema

jeremyh opened the pull request:
We know that all eo3 documents have a schema if they pass the validator tool, and all older datasets have no schema. So instead of heuristics, we can just use the `$schema` field to detect if it's eo3.

It also now rejects unknown schemas. Several of us have indexed eo3 documents accidentally recently, which ODC-1.7 quietly accepts despite not supporting eo3. It's friendlier to reject them immediately, and leaves room to match a Stac Item or Eo4 schema in the future.

- [x] Tests updated

jeremyh committed:
Simplify eo3 detection: use only $schema

We know that all eo3 documents have a schema if they pass the validator
tool, and all older datasets have no schema. So instead of heuristics,
we can just use the `$schema` field to detect if it's eo3.

It also now rejects unknown schemas. Several of us have indexed eo3
documents accidentally recently, which ODC-1.7 quietly accepts despite
not supporting eo3. It's friendlier to reject them immediately, and
leaves room to match a Stac Item or Eo4 schema in the future.

jeremyh committed:
Update test eo3 documents

Including required fields.

(currently geometry is also required if you have any other spatial fields, but specifying grid-only looks perhaps an exception we could add?)

Kirill888 commented:
Thanks @jeremyh, can I ask you to add a  little note to the document linked below regarding significance of the schema field:

https://github.com/opendatacube/datacube-core/blob/develop/docs/ops/dataset_documents.rst

jeremyh committed:
Expand eo3 docs

Update example doc slightly too:

- Use yaml timestamp type rather than strings
- Clarify that the 'name' is a product name, not platform.
- crs codes were lowercase by convention, following stac at the time of eo3's creation. It's a lint warning in eo3-validate, and other eodatasets3 outputs are lowercase.

jeremyh commented:
Yep -- added.

I'm biased, but I added a link to `eo3-validate` which I think will be useful for people too.
